toolchain/parse/state.def

```diff
@@ -227,63 +227,54 @@ CARBON_PARSE_STATE(CodeBlockFinish)
 
 // Handles a declaration name and parameters, such as `Foo[...](...)`.
 //
-// Allowed parameters:
-// - None: `Foo` only.
-// - Optional: `Foo`, `Foo(...)`, or `Foo[...](...)`.
-// - Required: `Foo(...)` or `Foo[...](...)`.
-//
 // name . ...
-// ^~~~
-//   1. PeriodAsDecl
-//   2. DeclNameAndParamsAfterNameAs(None|Optional|Required)
+// ^~~~~~
+//   1. DeclNameAndParams
 //
-// name ...
+// name [ ... ]
 // ^~~~
-//   1. DeclNameAndParamsAfterNameAs(None|Optional|Required)
-//
-//  ???
-// ^
-//   (state done)
-CARBON_PARSE_STATE_VARIANTS3(DeclNameAndParams, None, Optional, Required)
-
-// Handles a declaration name between the main name and implicit parameters.
-//
-// name . ...
-//     ^
-//   1. PeriodAsDecl
-//   2. DeclNameAndParamsAfterNameAs(None|Optional|Required)
-//
-// name [ ... ]    (variant is not None)
-//     ^
 //   1. PatternListAsImplicit
 //   2. DeclNameAndParamsAfterImplicit
 //
-// name ( ... )    (variant is not None)
-//     ^
+// name ( ... )
+// ^~~~
 //   1. PatternListAsTuple
+//   2. DeclNameAndParamsAfterParams
 //
-// name ...        (variant is not Required)
-//     ^
+// name ...
+// ^~~~
 //   (state done)
 //
-// name ???        (variant is Required)
-//     ^
+//  ???
+// ^
```

Comment:
What happened here?

Reply:
This is trying to say: if we're in the `DeclNameAndParams` state, and the first token isn't a name, that's an error and we're done. (We don't check there's a name next before pushing the state.)

Reply:
Maybe this comment could be made clearer?